vavr/src/main/java/io/vavr/control/Either.java

```diff
@@ -414,6 +414,44 @@ default void orElseRun(Consumer<? super L> action) {
         return isLeft() || predicate.test(get()) ? Option.some(this) : Option.none();
     }
 
+    /**
+     * Filters this right-biased {@code Either} by testing a predicate.
+     * If the {@code Either} is a Right and the predicate doesn't match, the
+     * {@code Either} will be turned into a {@code Left} with contents computed by applying
+     * the filterVal function to the {@code Either} value.
+     * <p>
+     *
+     * <pre>{@code
+     * import static io.vavr.API.*;
+     *
+     * class Example {
+     *
+     *     // = Left("bad: a")
+     *     Right("a").filterEither(i -> false, val -> "bad: " + val);
+     *
+     *     // = Right("a")
+     *     Right("a").filterEither(i -> true, val -> "bad: " + val);
+     *
+     * }
+     * }</pre>
+     *
+     * @param predicate A predicate
+     * @return an {@code Either} instance
+     * @throws NullPointerException if {@code predicate} is null
+     */
+    default Either<L,R> filterEither(Predicate<? super R> predicate,
+                                     Function<? super R, ? extends L> filterVal) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        Objects.requireNonNull(filterVal, "filterVal is null");
+        if (isLeft()) {
+            return this;
+        } else if (!predicate.test(get())) {
+            return Either.left(filterVal.apply(get()));
+        } else {
+            return this;
+        }
```

Comment:
This could be simplified to

```java
if (isLeft() || predicate.test(get())) {
    return this;
} else {
    return Either.left(filterVal.apply(get()));
}
```

or even

```java
return isLeft() || predicate.test(get()) ? this : left(zero.apply(get()));
```